kw6/stream.py

```diff
@@ -54,11 +84,13 @@ def seek_(self, frame_index: types.FRAME_INDEX):
             print('WARNING: stream position ahead of input seek position, rewinding to start of file.')
             self.stream.seek(settings.N_BYTES_VERSION)
 
-        while (
-            self.stream.peek(1) != b''
-            and PositionHeader.peek(self.stream).frame_index != frame_index
-        ):
-            Position.skip_(self.stream)
+        while not self.empty():
+            next_position = PositionHeader.peek(self.stream).frame_index
+            self.cached_positions[next_position] = self.stream.tell()
+            if next_position != frame_index:
+                Position.skip_(self.stream)
+            else:
+                break
 
     def _seek_closest_stored_position(self, frame_index: types.FRAME_INDEX):
```

Comment:
We save every position that we have seen right? Does it make sense to first try a simple lookup?

Reply:
the call to `self._seek_closest_stored_position` at the top of the method should cover this unless I'm mistaken?

Reply:
I meant a O(1) lookup in the dictionary before doing the for-loop in `self._seek_closest_stored_position`

Reply:
Ah sure, that's cleaner :) Done